common/lib/util/defaults.js

```diff
@@ -116,16 +127,22 @@ Defaults.normaliseOptions = function(options) {
 	if(!('queueMessages' in options))
 		options.queueMessages = true;
 
-	var production = false;
 	if(options.restHost) {
 		options.realtimeHost = options.realtimeHost || options.restHost;
+		options.fallbackHosts = options.fallbackHostsUseDefault && !options.port && !options.tlsPort ? Defaults.FALLBACK_HOSTS : options.fallbackHosts;
 	} else {
 		var environment = (options.environment && String(options.environment).toLowerCase()) || Defaults.ENVIRONMENT;
-		production = !environment || (environment === 'production');
+		var production = !environment || (environment === 'production');
 		options.restHost = production ? Defaults.REST_HOST : environment + '-' + Defaults.REST_HOST;
```

Comment:
@lmars maybe I am thinking too critical... is it possible for users to specify both rest and environment in options?
According to the current implementation, it will just ignore env. option. Not sure, if we want to generate an error for the same... like we did in fallbackhosts

Reply:
Just to put this in perspective: restHost and realtimeHost [are not actually documented client options](https://www.ably.io/documentation/realtime/usage#client-options). Customers should only be using `environment`, except for very rare exceptions like whitelabelled client libraries with enterprise customers, who we will give a set of client options to.

They're still part of the spec because having a standard way of pointing the library to a custom endpoint for testing is good for internal usability, but at the end of the day their usability is not very customer-facing.

Reply:
`restHost` and `environment` are mutually exclusive (see [`TO3k2`](https://www.ably.io/documentation/client-lib-development-guide/features#TO3k2)). I chose not to throw an error, as I think it's pretty clear to developers that `restHost` and `realtimeHost` take priority over `environment` anyway.